src/services/sectionService.ts

```diff
@@ -3,60 +3,126 @@ import { Section } from "../models/Section";
 import { Template } from "../models/Template";
 import { SectionTag } from "../models/SectionTag";
 import { Tag } from "../models/Tag";
+import { hasPermissionOnTemplate } from "./templateService";
 import { VersionedSection } from "../models/VersionedSection";
-import { ForbiddenError, NotFoundError } from "../utils/graphQLErrors";
+import { NotFoundError } from "../utils/graphQLErrors";
+import { Question } from "../models/Question";
+import { generateQuestionVersion } from "./questionService";
+import { formatLogMessage, logger } from "../logger";
 
+// Creates a new Version/Snapshot the specified Section (as a point in time snapshot)
+//    - Creates a new VersionedSection including all of the related Questions
+//    - Resets the isDirty flag on the Section
+export const generateSectionVersion = async (
+  context: MyContext,
+  section: Section,
+  versionedTemplateId: number,
+): Promise<boolean> => {
+
+  // If the section has no id then it has not yet been saved so throw an error
+  if (!section.id) {
+    throw new Error('Cannot publish unsaved Section');
+  }
+
+  // Create the new Version
+  const versionedSection = new VersionedSection({
+    versionedTemplateId: versionedTemplateId,
+    sectionId: section.id,
+    name: section.name,
+    introduction: section.introduction,
+    requirements: section.requirements,
+    guidance: section.guidance,
+    displayOrder: section.displayOrder,
+    tags: section.tags,
+  });
+
+  try {
+    const created = await versionedSection.create(context);
+
+    // If the creation was successful
+    if (created && (!created.errors || (Array.isArray(created.errors) && created.errors.length === 0))) {
+      // Create a version for all the associated questions
+      const questions = await Question.findBySectionId('generateSectionVersion', context, section.id);
+      let allQuestionsWereVersioned = true;
+
+      questions.forEach(async (question) => {
+        const questionInstance = new Question({
```

Comment:
Need to create an instance of Question here because "generateQuestionVersion" tries to call .create() on question, so it needs to be a proper instance of it.